zone/expedition_request.cpp

```diff
@@ -0,0 +1,393 @@
+	m_leader_id = m_leader ? m_leader->CharacterID() : database.GetCharacterID(m_leader_name.c_str());
+
+	std::vector<std::string> member_names;
+	member_names.emplace_back(m_leader_name); // leader always added first
+
+	for (int i = 0; i < MAX_GROUP_MEMBERS; ++i)
+	{
+		if (group->membername[i][0] && m_leader_name != group->membername[i])
+		{
+			member_names.emplace_back(group->membername[i]);
+		}
+	}
+
+	if (member_names.size() > m_max_players)
+	{
+		m_not_all_added_msg = fmt::format(CREATE_NOT_ALL_ADDED, "group", SystemName,
+			SystemName, m_max_players, "group", member_names.size());
+	}
+
+	return CanMembersJoin(member_names);
+}
+
+std::string ExpeditionRequest::GetGroupLeaderName(uint32_t group_id)
+{
+	char leader_name_buffer[64] = { 0 };
+	database.GetGroupLeadershipInfo(group_id, leader_name_buffer);
+	return std::string(leader_name_buffer);
+}
+
+bool ExpeditionRequest::CanMembersJoin(const std::vector<std::string>& member_names)
+{
+	if (member_names.empty())
+	{
+		return false;
+	}
+
+	bool requirements_met = true;
+
+	if (CheckMembersForConflicts(member_names))
+	{
+		requirements_met = false;
+	}
+
+	// live only checks player count requirement after other expensive checks pass (?)
+	// maybe it's done intentionally as a way to preview lockout conflicts
+	if (requirements_met)
+	{
+		requirements_met = IsPlayerCountValidated();
+	}
+
+	return requirements_met;
+}
+
+bool ExpeditionRequest::LoadLeaderLockouts()
+{
+	// leader's lockouts are used to check member conflicts and later stored in expedition
+	auto lockouts = ExpeditionDatabase::LoadCharacterLockouts(m_leader_id, m_expedition_name);
+
+	for (auto& lockout : lockouts)
+	{
+		if (!lockout.IsExpired())
+		{
+			m_lockouts.emplace(lockout.GetEventName(), lockout);
+
+			// on live if leader has a replay lockout it never bothers checking for event conflicts
+			if (m_check_event_lockouts && lockout.IsReplayTimer())
+			{
+				m_check_event_lockouts = false;
+			}
+		}
+	}
+
+	return true;
+}
+
+bool ExpeditionRequest::CheckMembersForConflicts(const std::vector<std::string>& member_names)
+{
+	// load data for each member and compare with leader lockouts
+	auto results = ExpeditionDatabase::LoadMembersForCreateRequest(member_names, m_expedition_name);
+	if (!results.Success() || !LoadLeaderLockouts())
+	{
+		LogExpeditions("Failed to load data to verify members for expedition request");
+		return true;
+	}
+
+	bool is_solo = (member_names.size() == 1);
+	bool has_conflicts = false;
+
+	std::vector<ExpeditionRequestConflict> member_lockout_conflicts;
+
+	uint32_t last_character_id = 0;
+	for (auto row = results.begin(); row != results.end(); ++row)
+	{
+		auto character_id = static_cast<uint32_t>(std::strtoul(row[0], nullptr, 10));
+		std::string character_name(row[1]);
+		bool has_expedition = (row[2] != nullptr); // in expedition_members with another expedition
```

Comment:
Can we define these as variables so this is a bit more readable; you kind of need to magically know what column is what when looking at this